plutus-core/plutus-ir/src/PlutusIR/Compiler/Datatype.hs

```diff
@@ -342,6 +444,40 @@ mkConstructor dty d@(Datatype ann _ tvs _ constrs) index = do
         argNames <- for [0..(length argTypes -1)] (\i -> safeFreshName $ "arg_" <> showText i)
         pure $ zipWith (VarDecl ann) argNames argTypes
 
+    constrBody <- case _dcoStyle opts of
+          SumsOfProducts -> do
+            -- We have to be a bit careful annotating the type of the constr. It is inside the 'wrap' so it
+            -- needs to be one level "unrolled".
+
+            -- The pattern functor with a hole in it
+            pf <- mkDatatypePatternFunctor opts ann d
+            -- ... and with the hole filled in with the datatype type, by making
+            -- a type lambda and immediately applying it
+            let unrolled = TyApp ann (PIR.mkIterTyLam [tv] pf) (getType dty)
```

Comment:
@effectfully I think this might be wrong, or at least suspicious. I think this is exactly the type of the term "inside" a `wrap`. I don't know if we have a way to get that with the current recursive types machinery or whether it would be useful to add one?

Reply:
No, it's fine. Roughly, inside a `wrap` we expect the term to have this complex type which is the pattern functor applied to the ifixed version of the pattern functor i.e. "unrolled" one time. That's exactly the type we need here, but we need to actually write it down on a term, which we don't normally, and it's annoying. And I'm suspicious I might do it wrong, hence why I'm wondering if the machinery might do it for us.

Reply:
> @effectfully I think this might be wrong, or at least suspicious. I think this is exactly the type of the term "inside" a `wrap`. I don't know if we have a way to get that with the current recursive types machinery or whether it would be useful to add one?

I've spent hours staring at that single line. I'm going to take look at it from a different perspective.

And I'm still kinda astonished that it's possible to make `ifix` interoperate with sums-of-products in such an extremely concise and non-intrusive way.